gdal/alg/marching_squares/square.h

```diff
@@ -0,0 +1,425 @@
+        case (ALL_HIGH & ~LOWER_LEFT):
+            //debug("ALL_HIGH & ~LOWER_LEFT");
+            return Segments(Segment(interpolate(LOWER_BORDER, level), interpolate(LEFT_BORDER, level)));
+        case (ALL_HIGH & ~LOWER_RIGHT):
+            //debug("ALL_HIGH & ~LOWER_RIGHT");
+            return Segments(Segment(interpolate(RIGHT_BORDER, level), interpolate(LOWER_BORDER, level)));
+        case (ALL_HIGH & ~UPPER_RIGHT):
+            //debug("ALL_HIGH & ~UPPER_RIGHT");
+            return Segments(Segment(interpolate(UPPER_BORDER, level), interpolate(RIGHT_BORDER, level)));
+        case (SADDLE_NE):
+            //debug("SADDLE_NE");
+            return (fudge(level, center().value) > level)
+                ? Segments(
+                    Segment(interpolate(UPPER_BORDER, level), interpolate(LEFT_BORDER, level)),
+                    Segment(interpolate(LOWER_BORDER, level), interpolate(RIGHT_BORDER, level)))
+                : Segments(
+                    Segment(interpolate(LOWER_BORDER, level), interpolate(LEFT_BORDER, level)),
+                    Segment(interpolate(UPPER_BORDER, level), interpolate(RIGHT_BORDER, level)));
+        case (SADDLE_NW):
+            //debug("SADDLE_NW");
+            return (fudge(level, center().value) > level)
+                ? Segments(
+                    Segment(interpolate(LEFT_BORDER, level), interpolate(LOWER_BORDER, level)),
+                    Segment(interpolate(RIGHT_BORDER, level), interpolate(UPPER_BORDER, level)))
+                : Segments(
+                    Segment(interpolate(RIGHT_BORDER, level), interpolate(LOWER_BORDER, level)),
+                    Segment(interpolate(LEFT_BORDER, level), interpolate(UPPER_BORDER, level)));
+        }
+        assert(false);
+        return Segments();
+    }
+
+    template <typename Writer, typename LevelGenerator>
+    void process(const LevelGenerator &levelGenerator, Writer & writer) const
+    {
+        if (nanCount == 4) // nothing to do
+            return;
+
+
+        if (nanCount) // split in 4
+        {
+            if (!isnan(upperLeft.value))
+                upperLeftSquare().process(levelGenerator, writer);
+            if (!isnan(upperRight.value))
+                upperRightSquare().process(levelGenerator, writer);
+            if (!isnan(lowerLeft.value))
+                lowerLeftSquare().process(levelGenerator, writer);
+            if (!isnan(lowerRight.value))
+                lowerRightSquare().process(levelGenerator, writer);
+            return;
+        }
+
+        if ( writer.polygonize && borders )
+        {
+            const std::size_t nbBorders = 4;
+            const unsigned char squareBorders[nbBorders] = {UPPER_BORDER, LEFT_BORDER, RIGHT_BORDER, LOWER_BORDER};
+            for (const unsigned char *brd=squareBorders; brd != squareBorders + nbBorders; brd++)
+            {
+                // if we have this border on the square
+                if (*brd & borders)
```

Comment:
is this condition needed ? borders alreay tested for nullptr before, and *brd cannot be zero given the constatn values

Reply:
This is a bitwise and here. I will try to make it more readable :)

Reply:
ah sorry missed that. I guess if ( (*brd & borders) != 0 ) could be less ambiguous